File: be/src/main/java/sg/flow/services/AccountServices/AccountServiceImpl.java
package sg.flow.services.AccountServices;

import java.util.List;
import java.util.Optional;

import org.springframework.stereotype.Service;

import lombok.RequiredArgsConstructor;
import sg.flow.entities.Account;
import sg.flow.models.account.AccountWithTransactionHistory;
import sg.flow.models.account.BriefAccount;
import sg.flow.models.transaction.TransactionHistoryDetail;
import sg.flow.repositories.account.AccountRepository;
import sg.flow.repositories.transactionHistory.TransactionHistoryRepository;

@Service
@RequiredArgsConstructor
public class AccountServiceImpl implements AccountService {

    private final AccountRepository accountRepository;

    private final TransactionHistoryRepository transactionHistoryRepository;

    @Override
    public List<BriefAccount> getBriefAccounts(Integer userId) {
        return accountRepository.findBriefAccountsOfUser(userId);
    }

    @Override
    public List<AccountWithTransactionHistory> getAccountWithTransactionHistory(Integer userId) {
        return accountRepository.findAccountWithTransactionHistorysOfUser(userId);
    }

    @Override
    public BriefAccount getBriefAccount(Integer userId, Long accountId) throws IllegalArgumentException {
        Optional<Account> accountFound = accountRepository.findById(accountId);
        if (accountFound.isPresent()) {
            Account account = accountFound.get();
            if (account.getOwner().getId().equals(userId)) {
                return BriefAccount.builder()
                        .id(account.getId())
                        .bank(account.getBank())
                        .balance(account.getBalance())
                        .accountName(account.getAccountName())
                        .build();
            } else {
                throw new IllegalArgumentException("Account does not belong to user");
            }
        }
        return null;
    }

    @Override
    public AccountWithTransactionHistory getAccountWithTransactionHistory(Integer userId, Long accountId) throws IllegalArgumentException {
        Optional<Account> accountFound = accountRepository.findById(accountId);
        if (accountFound.isPresent()) {
            Account account = accountFound.get();
            if (account.getOwner().getId().equals(userId)) {
                AccountWithTransactionHistory accountWithTransactionHistory = AccountWithTransactionHistory.builder()
                        .id(account.getId())
                        .accountNumber(account.getAccountNumber())
                        .bank(account.getBank())
                        .balance(account.getBalance())
                        .accountName(account.getAccountName())
                        .accountType(account.getAccountType())
                        .interestRatePerAnnum(account.getInterestRatePerAnnum())
                        .build();
                
                List<TransactionHistoryDetail> recentTransactionDetails = transactionHistoryRepository.findRecentTransactionHistoryDetailOfAccount(accountId);
                System.out.println(recentTransactionDetails);
                accountWithTransactionHistory.setRecentTransactionHistoryDetails(recentTransactionDetails);

                return accountWithTransactionHistory;
            } else {
                throw new IllegalArgumentException("Account does not belong to user");
            }
        }
        return null;
    }

}

File: be/src/main/kotlin/sg/flow/services/AccountServices/AccountServiceImpl.kt
package sg.flow.services.AccountServices

import org.springframework.stereotype.Service
import sg.flow.entities.utils.AccountType
import sg.flow.grpc.exception.AccountDoesNotExistException
import sg.flow.grpc.exception.RequestedAccountNotBelongException
import sg.flow.models.account.AccountWithTransactionHistory
import sg.flow.models.account.BriefAccount as Account
import sg.flow.repositories.account.AccountRepository
import sg.flow.repositories.transactionHistory.TransactionHistoryRepository

@Service
class AccountServiceImpl(
        private val accountRepository: AccountRepository,
        private val transactionHistoryRepository: TransactionHistoryRepository,
) : AccountService {

        override suspend fun getAccounts(userId: Int): List<Account> {
                val accounts = accountRepository.findAccountsOfUser(userId)

                fun isNotCard(accountTypeStr: String): Boolean {
                        val accountType = AccountType.fromValue(accountTypeStr)
                        return accountType != AccountType.DEBIT_CARD && accountType != AccountType.CREDIT_CARD
                }

                val accountsWithOutCards = accounts.filter { account -> isNotCard(account.accountType) }
                return accountsWithOutCards
        }


        override suspend fun getAccountWithTransactionHistory(
                userId: Int
        ): List<AccountWithTransactionHistory> =
                accountRepository.findAccountWithTransactionHistorysOfUser(userId)

        override suspend fun getAccount(userId: Int, accountId: Long): Account {
                val account =
                        accountRepository.findById(accountId)
                                ?: throw AccountDoesNotExistException(accountId)

                if (account.owner.id != userId) {
                        throw RequestedAccountNotBelongException(userId, accountId)
                }

                return Account(
                        id = account.id ?: -1,
                        bank = account.bank,
                        balance = account.balance,
                        accountName = account.accountName,
                        accountType = account.accountType.toString(),
                        accountNumber = account.accountNumber

                )
        }

        override suspend fun getAccountWithTransactionHistory(
                userId: Int,
                accountId: Long
        ): AccountWithTransactionHistory {
                val account =
                        accountRepository.findById(accountId)
                                ?: throw AccountDoesNotExistException(accountId)

                if (account.owner.id != userId) {
                        throw RequestedAccountNotBelongException(userId, accountId)
                }

                val recentTransactionDetails =
                        transactionHistoryRepository.findRecentTransactionHistoryDetailOfAccount(
                                accountId
                        )

                return AccountWithTransactionHistory(
                        id = account.id ?: -1,
                        accountNumber = account.accountNumber,
                        bank = account.bank,
                        balance = account.balance,
                        accountName = account.accountName,
                        accountType = account.accountType,
                        interestRatePerAnnum = account.interestRatePerAnnum,
                        recentTransactionHistoryDetails = recentTransactionDetails
                )
        }
}
